auth/rest.go

```diff
@@ -114,6 +121,82 @@ func (a *RestAuthDecorator) DecorateHandler(orig httprouter.Handle, appName stri
 			}
 
 			orig(res, req, p)
+
+			// if app was a provider app allow token rewrites
+			if cfg.Authentication.ProviderConfig.Service == appName {
+				// rewrite body tokens
+				bodyTokenKey := res.Header().Get("X-Gateway-BodyToken")
+				if bodyTokenKey != "" {
+
+					var response map[string]string
+					jsonBlob, _ := ioutil.ReadAll(req.Body)
```

Comment:
Unsure here because a responsewriter has no body attribute and i'm not sure how to access the response from orig(res, req, p)

Reply:
OK, this won't work, as you're reading from the _request_ body, and not the _response_ body. Also, the `ResponseWriter` does not contain a body attribute, since it's not required to _record_ any data written to it (by default, it will probably stream right into the network connection without buffering). You might be able to replace the `ResponseWriter` passed into `orig` with a [ResponseRecorder](https://golang.org/pkg/net/http/httptest/#ResponseRecorder), which also implements the `ResponseWriter` interface, but also has a body that can be read.

Reply:
Also, please check the error returned by `ioutil.ReadAll`.